### PR TITLE
Ensure that DatetimeArray keeps reference to original data

### PR DESCRIPTION
**Pull request by TomAugspurger**
(cherry picked from commit 49f6495ba5ceef35a7a962f20e8b863c544592b7)

This avoids a segfault I (and possibly @jbrockmendel) were seeing on our datetimearray branches from us freeing a region of memory we didn't allocate.

I haven't really groked things yet, but `_libs/reduction.pyx::Slider` is doing some strange things with the data of the array passed to it (overwriting the buffer's `.data` and `.strides` attributes). One of the arguments, `buff`, is a slice on the original index. It *seems* like ensuring that the ndarray backing the DatetimeArray sliced index is the same as the original index avoids the segfault.

### CODE REVIEW

File: pandas/_libs/tslibs/conversion.pyx [hunk 1]
```diff
@@ -99,11 +99,11 @@ def ensure_datetime64ns(arr: ndarray, copy: bool=True):
 
     ivalues = arr.view(np.int64).ravel()
 
-    result = np.empty(shape, dtype='M8[ns]')
+    result = np.empty(shape, dtype=NS_DTYPE)
     iresult = result.ravel().view(np.int64)
 
     if len(iresult) == 0:
-        return result
+        return arr.astype(NS_DTYPE, copy=copy)
```
Comment (jbrockmendel) [inline pandas/_libs/tslibs/conversion.pyx]: why is this change needed?

Comment (jbrockmendel) [reply pandas/_libs/tslibs/conversion.pyx]: never mind, addressed in OP

Comment (jreback) [reply pandas/_libs/tslibs/conversion.pyx]: I think this should actually be ``arr.view(NS_DTYPE)``

**Review by jreback — CHANGES_REQUESTED** [review]

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks!
